File: backend/app/schemas/mcp.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class ConnectionCreate(BaseModel):
    server_type: Literal["analytics", "file"]
    name: str | None = Field(None, min_length=1, max_length=120)
    model_config = ConfigDict(extra="forbid")


class ConnectionOut(BaseModel):
    id: UUID; name: str; server_type: str; transport: str; status: str; is_enabled: bool; created_at: datetime; updated_at: datetime


class ToolUpdate(BaseModel):
    is_enabled: bool | None = None
    requires_approval: bool | None = None
    model_config = ConfigDict(extra="forbid")


class ToolOut(BaseModel):
    id: UUID; connection_id: UUID; connection_name: str; external_name: str; display_name: str
    description: str | None; input_schema: dict[str, Any]; is_enabled: bool
    requires_approval: bool; risk_level: str; discovered_at: datetime; updated_at: datetime


class ToolExecute(BaseModel):
    arguments: dict[str, Any] = Field(default_factory=dict)
    model_config = ConfigDict(extra="forbid")


class ToolCallOut(BaseModel):
    id: UUID; conversation_id: UUID | None; message_id: UUID | None; mcp_tool_id: UUID
    tool_name: str; status: str; arguments_summary: dict[str, Any]
    result_summary: str | None; error_message: str | None; started_at: datetime
    completed_at: datetime | None; created_at: datetime; approval_id: UUID | None = None
    final_message_content: str | None = None
    model_config = ConfigDict(from_attributes=True)


class ApprovalOut(BaseModel):
    id: UUID; tool_call_id: UUID; tool_name: str; status: str; risk_level: str
    arguments_summary: dict[str, Any]; requested_at: datetime; resolved_at: datetime | None

File: frontend/src/types/mcp.ts
export type MCPConnection = {
  id: string; name: string; server_type: "analytics" | "file"; transport: string;
  status: "connected" | "disconnected" | "error"; is_enabled: boolean;
  created_at: string; updated_at: string;
};
export type MCPTool = {
  id: string; connection_id: string; connection_name: string; external_name: string;
  display_name: string; description: string | null; input_schema: Record<string, unknown>;
  is_enabled: boolean; requires_approval: boolean; risk_level: string;
  discovered_at: string; updated_at: string;
};
export type ToolCall = {
  id: string; conversation_id?: string | null; message_id?: string | null; mcp_tool_id?: string;
  tool_name: string; status: "pending" | "awaiting_approval" | "running" | "completed" | "failed" | "denied";
  arguments_summary: Record<string, unknown>; result_summary: string | null; error_message?: string | null;
  risk_level?: string; approval_id?: string | null; started_at?: string; completed_at?: string | null; created_at?: string;
  final_message_content?: string | null;
};
export type Approval = {
  id: string; tool_call_id: string; tool_name: string; status: string; risk_level: string;
  arguments_summary: Record<string, unknown>; requested_at: string; resolved_at: string | null;
};
